src/gui/owncloudgui.cpp

```diff
@@ -173,11 +173,11 @@ void ownCloudGui::slotTrayClicked(QSystemTrayIcon::ActivationReason reason)
     if (reason == QSystemTrayIcon::Trigger) {
         if (OwncloudSetupWizard::bringWizardToFrontIfVisible()) {
             // brought wizard to front
-        } else if (_shareDialogs.size() > 0) {
-            // Share dialog(s) be hidden by other apps, bring them back
-            Q_FOREACH (const QPointer<ShareDialog> &shareDialog, _shareDialogs) {
-                Q_ASSERT(shareDialog.data());
-                raiseDialog(shareDialog);
+        } else if (!_visibleDialogs.isEmpty()) {
```

Comment:
Now that makes me concerned... before the only windows raised were the share dialogs... but now that'll be a much more common case since that could be any dialog including the settings one.

In turn that means that as long as any window is open I can't reach the tray main dialog anymore. And it's kind of common to go back and forth between the settings dialog and the tray dialog... but the latter auto-closes when it looses the focus and clicking the systray icon won't bring it back anymore!

At the same time I'm not sure we want everything to be brought to the front every time that tray window is shown because we clicked on the tray icon.

I generally think messing with the window states too much is unwise because at some point it means you're trying to do the window manager job and you're bound to break behaviors people are used to outside of your app.

@jancborchardt any opinion on my ramblings there?

Reply:
I'm sharing the same concerns to some extent. My main motivation for the whole PR: The dialogs, like share and re-auth are loose, so there currently is no way to bring them back. They don't get any task bar entry and especially on macOS they are completely untracked. Users have to minimize or close all other open windows to get them back and re-auth dialogs often will remain unnoticed.

Example: Re-auth dialog opens browser and gets hidden behind. User authenticates and then clicks the Tray icon, what should bring the re-auth dialog back to front.

Note that right-clicking the Tray still opens the context menu, allowing to reach the main dialog / settings anyway. But of course we may also decide to change above code to raise the dialogs AND open the Tray menu.

Reply:
> I'm sharing the same concerns to some extent. My main motivation for the whole PR: The dialogs, like share and re-auth are loose, so there currently is no way to bring them back. They don't get any task bar entry and especially on macOS they are completely untracked. Users have to minimize or close all other open windows to get them back and re-auth dialogs often will remain unnoticed.
> 
> Example: Re-auth dialog opens browser and gets hidden behind. User authenticates and then clicks the Tray icon, what should bring the re-auth dialog back to front.

Yep, I'm not saying there is no need... I'm just concerned about the side effects. Just want to make sure they are properly weighted.

> Note that right-clicking the Tray still opens the context menu, allowing to reach the main dialog / settings anyway.

Except... not on Mac. Both right and left click are supposed to do the same thing there (if not that's a bug and wasn't intentional).

> But of course we may also decide to change above code to raise the dialogs AND open the Tray menu.

Yep that's something I was alluding at in my previous comment but now that means everything is always brought forward when the tray menu is opened. Unsure it's better... well at least you can reach everything but that feels a bit thrown in your face all the time. :-)

Reply:
Good question for sure. :\ As usual: Did you check how Dropbox and Google Drive et al handle it?

(Not sure if this is the thing that is being talked about, but what I’ve been noticing on Ubuntu, is that when I want to share a file with the "Share options", that dialog doesn’t pop up but stay in the background and I need to Alt-Tab to get it.)

Reply:
@er-vin :
> Yep, I'm not saying there is no need... I'm just concerned about the side effects. Just want to make sure they are properly weighted.

I know, just wanted to make the idea as clear as possible, can be confusing ;-)
And I agree of course, to make sure they are properly weighted. Happy for more ideas how to address the problem :)

> 
> > Note that right-clicking the Tray still opens the context menu, allowing to reach the main dialog / settings anyway.
> 
> Except... not on Mac. Both right and left click are supposed to do the same thing there (if not that's a bug and wasn't intentional).

That's what the depending #2155 is for, the Dock icon is macOS's native context menu handler and serves the context menu ;)

> 
> > But of course we may also decide to change above code to raise the dialogs AND open the Tray menu.
> 
> Yep that's something I was alluding at in my previous comment but now that means everything is always brought forward when the tray menu is opened. Unsure it's better... well at least you can reach everything but that feels a bit thrown in your face all the time. :-)

Yeah, exactly ^^

Reply:
@jancborchardt :
> Good question for sure. :\ As usual: Did you check how Dropbox and Google Drive et al handle it?

Yeah, right? This thing made my brain spinning back and forth over the last couple of months ;-/

Didn't check Dropbox or Google Drive.

> 
> (Not sure if this is the thing that is being talked about, but what I’ve been noticing on Ubuntu, is that when I want to share a file with the "Share options", that dialog doesn’t pop up but stay in the background and I need to Alt-Tab to get it.)

And on macOS it's even worse because Alt-Tab (Cmd-Tab on Mac) wouldn't work, also not for the Settings dialog. It only works when the Dock icon is visible (#2155) AND we bring the dialogs to front on Dock-click :/